File: bindings/ruby/vibe.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# libvibe — Ruby binding using Fiddle (Ruby standard library, no gems).
#
#   require_relative "vibe"
#   doc = Vibe.parse(File.binread("config.vibe"))
#   host = doc.get_string("server.host")

require "fiddle"
require "fiddle/import"

module Vibe
  module LibC
    extend Fiddle::Importer
    lib = ENV["VIBE_LIB"] || File.expand_path("../../libvibe.so", __dir__)
    dlload lib
    extern "const char* vibe_version()"
    extern "void* vibe_parse(const char*, size_t, void*)"
    extern "const char* vibe_get_string(void*, const char*)"
    extern "long long vibe_get_int(void*, const char*)"
    extern "double vibe_get_float(void*, const char*)"
    extern "char vibe_get_bool(void*, const char*)"
    extern "void* vibe_get_array(void*, const char*)"
    extern "size_t vibe_array_size(void*)"
    extern "char* vibe_emit(void*)"
    extern "void vibe_free(void*)"
    extern "void vibe_value_free(void*)"
    extern "const char* vibe_error_code_string(int)"
  end

  class Error < StandardError
    attr_reader :code, :line, :column
    def initialize(code, message, line, column)
      @code = code
      @line = line
      @column = column
      super("#{message} (code #{code} at #{line}:#{column})")
    end
  end

  def self.version
    LibC.vibe_version.to_s
  end

  class Doc
    def initialize(ptr)
      @ptr = ptr
    end

    def get_string(path)
      p = LibC.vibe_get_string(@ptr, path)
      p.null? ? nil : p.to_s
    end

    def get_int(path)  = LibC.vibe_get_int(@ptr, path)
    def get_float(path) = LibC.vibe_get_float(@ptr, path)
    def get_bool(path) = LibC.vibe_get_bool(@ptr, path) != 0

    def array_size(path)
      arr = LibC.vibe_get_array(@ptr, path)
      arr.null? ? 0 : LibC.vibe_array_size(arr)
    end

    def emit
      raw = LibC.vibe_emit(@ptr)
      return "" if raw.null?
      s = raw.to_s
      LibC.vibe_free(raw)
      s
    end

    def free
      LibC.vibe_value_free(@ptr) unless @ptr.null?
      @ptr = Fiddle::NULL
    end
  end

  # VibeError { bool@0; int code@4; char* message@8; int line@16; int column@20 } = 24 bytes
  def self.parse(data)
    data = data.b
    buf = ("\0".b * 24)
    err = Fiddle::Pointer[buf]
    ptr = LibC.vibe_parse(data, data.bytesize, err)
    return Doc.new(ptr) unless ptr.null?

    code_num = buf[4, 4].unpack1("l")
    msg_addr = buf[8, 8].unpack1("J")
    line = buf[16, 4].unpack1("l")
    col  = buf[20, 4].unpack1("l")
    code = LibC.vibe_error_code_string(code_num).to_s
    msg  = msg_addr.zero? ? "parse error" : Fiddle::Pointer.new(msg_addr).to_s
    raise Error.new(code, msg, line, col)
  end
end

if __FILE__ == $PROGRAM_NAME
  sample = ENV["VIBE_SAMPLE"] || File.expand_path("../sample.vibe", __dir__)
  doc = Vibe.parse(File.binread(sample))
  ok = true
  checks = [
    ["version",     Vibe.version,              "1.1.0"],
    ["name",        doc.get_string("name"),    "libvibe"],
    ["answer",      doc.get_int("answer"),     42],
    ["pi",          doc.get_float("pi").round(5), 3.14159],
    ["enabled",     doc.get_bool("enabled"),   true],
    ["server.host", doc.get_string("server.host"), "localhost"],
    ["server.port", doc.get_int("server.port"), 8080],
    ["len(ports)",  doc.array_size("ports"),   3],
  ]
  checks.each do |name, got, want|
    ok = false if got != want
    puts "  [#{got == want ? 'ok ' : 'BAD'}] #{name} = #{got.inspect}"
  end
  if doc.emit.include?("libvibe")
    puts "  [ok ] emit() round-trips"
  else
    ok = false; puts "  [BAD] emit() did not round-trip"
  end
  begin
    Vibe.parse("name {")
    ok = false; puts "  [BAD] malformed input did not raise"
  rescue Vibe::Error => e
    puts "  [ok ] rejects malformed input: #{e.code}"
  end
  puts(ok ? "ALL OK (ruby)" : "FAILED (ruby)")
  exit(ok ? 0 : 1)
end

File: bindings/lua/vibe.lua
-- libvibe — LuaJIT binding using the built-in FFI (no extra modules).
--
--   local vibe = require("vibe")
--   local doc = vibe.parse(io.open("config.vibe", "rb"):read("*a"))
--   print(doc:get_string("server.host"))
--
-- Requires LuaJIT (for the `ffi` library). Plain PUC Lua has no FFI.

local ffi = require("ffi")

ffi.cdef([[
  typedef struct { bool has_error; int code; char* message; int line; int column; } VibeError;
  const char* vibe_version(void);
  void*       vibe_parse(const char*, size_t, VibeError*);
  const char* vibe_get_string(void*, const char*);
  int64_t     vibe_get_int(void*, const char*);
  double      vibe_get_float(void*, const char*);
  bool        vibe_get_bool(void*, const char*);
  void*       vibe_get_array(void*, const char*);
  size_t      vibe_array_size(void*);
  char*       vibe_emit(void*);
  void        vibe_free(void*);
  void        vibe_value_free(void*);
  const char* vibe_error_code_string(int);
]])

local L = ffi.load(os.getenv("VIBE_LIB") or (arg and arg[0]:gsub("[^/]*$", "") or "") .. "../../libvibe.so")

local vibe = {}

function vibe.version()
  return ffi.string(L.vibe_version())
end

local Doc = {}
Doc.__index = Doc

function Doc:get_string(path)
  local p = L.vibe_get_string(self.ptr, path)
  return p ~= nil and ffi.string(p) or nil
end

function Doc:get_int(path)   return tonumber(L.vibe_get_int(self.ptr, path)) end
function Doc:get_float(path) return L.vibe_get_float(self.ptr, path) end
function Doc:get_bool(path)  return L.vibe_get_bool(self.ptr, path) end

function Doc:array_size(path)
  local arr = L.vibe_get_array(self.ptr, path)
  return arr ~= nil and tonumber(L.vibe_array_size(arr)) or 0
end

function Doc:emit()
  local raw = L.vibe_emit(self.ptr)
  if raw == nil then return "" end
  local s = ffi.string(raw)
  L.vibe_free(raw)
  return s
end

function Doc:free()
  if self.ptr ~= nil then L.vibe_value_free(self.ptr); self.ptr = nil end
end

function vibe.parse(data)
  local err = ffi.new("VibeError")
  local ptr = L.vibe_parse(data, #data, err)
  if ptr == nil then
    local code = ffi.string(L.vibe_error_code_string(err.code))
    local msg = err.message ~= nil and ffi.string(err.message) or "parse error"
    error(("VibeError: %s (%s at %d:%d)"):format(msg, code, err.line, err.column), 0)
  end
  return setmetatable({ ptr = ptr }, Doc)
end

-- self-test ----------------------------------------------------------------
if arg and (arg[0] or ""):match("vibe%.lua$") then
  local sample = os.getenv("VIBE_SAMPLE") or (arg[0]:gsub("[^/]*$", "") .. "../sample.vibe")
  local f = assert(io.open(sample, "rb"))
  local doc = vibe.parse(f:read("*a"))
  f:close()
  local ok = true
  local function check(name, got, want)
    if got ~= want then ok = false end
    print(("  [%s] %s = %s"):format(got == want and "ok " or "BAD", name, tostring(got)))
  end
  check("version", vibe.version(), "1.2.0")
  check("name", doc:get_string("name"), "libvibe")
  check("answer", doc:get_int("answer"), 42)
  check("pi", tonumber(string.format("%.5f", doc:get_float("pi"))), 3.14159)
  check("enabled", doc:get_bool("enabled"), true)
  check("server.host", doc:get_string("server.host"), "localhost")
  check("server.port", doc:get_int("server.port"), 8080)
  check("len(ports)", doc:array_size("ports"), 3)
  if doc:emit():find("libvibe", 1, true) then
    print("  [ok ] emit() round-trips")
  else ok = false; print("  [BAD] emit() did not round-trip") end
  local caught = pcall(function() vibe.parse("name {") end)
  if caught then ok = false; print("  [BAD] malformed input did not raise")
  else print("  [ok ] rejects malformed input") end
  print(ok and "ALL OK (luajit)" or "FAILED (luajit)")
  os.exit(ok and 0 or 1)
end

return vibe
